### ISSUE TITLE
Gulikit KK3 Max controller experiences ghost presses of the Home/Steam button

### ISSUE THREAD

**Issue by Soundtoxin** [ValveSoftware/SteamOS#1451]
#### Your system information

* Steam client version: 1710458266
* SteamOS version: 3.5.17
* Opted into Steam client beta?: [No] 
* Opted into SteamOS beta?: [No] 
* Have you checked for updates in Settings > System?: [Yes]

#### Please describe your issue in as much detail as possible:
Sitting in menus or playing games, every so often the Steam menu pops open by itself, even if the controller isn't in my hand it can happen. This interrupts games and is very annoying. Also it can activate various shortcuts depending on what buttons or sticks I'm holding at time of activation, such as moving the mouse briefly if I was playing a twin stick shooter using the right stick at the time of the issue. I should never have unintended erroneous inputs of this sort.

This issue is not confirmed to happen in any mode but XInput Mode, not confirmed to happen when using bluetooth or the included wireless receiver, happens with multiple cables and docks/hubs, does *not* happen with other controllers using the same cables and docks/hubs. I am using USB A to USB C cables. I am *not* using the cable that came with the controller, just various cables I already had around. Did not want to unwrap and wear the cable unnecessarily. Updating the controller's firmware to the latest version neither helped nor hurt.

I have already returned the controller and re-bought it once after extensive testing and discussion with GuliKit customer service had convinced me it was probably a hardware defect, however my replacement having the exact same issue right away made me question if it was really a hardware issue. I chose the gray color with purple buttons both times. I bought from Amazon both times. First from what looked like the official GuliKit store, later through the more popular AKNES listing (I believe they're a middle-man/reseller).

I also have a Reddit thread on /r/controllers where I detailed some of my issues and testing in the beginning. In that thread one other person mentioned ghost presses of the DPad and the B button using Windows, they returned it and did not get a replacement. A different person mentioned having issues playing with their *Steam Link* directly wired, but not with Windows or when playing wirelessly.

https://www.reddit.com/r/Controller/comments/1bgp8ou/gulikit_kk3_max_ghost_presses_on_steam_deck/

I do not know for sure that this is an issue with SteamOS, but if someone else could help me get to the bottom of what's going on here, that would be amazing. I had planned to use this controller almost exclusively wired, as that is my preference and I already have cables run and left plugged in at my desk for easily using various controllers, so the fact that it doesn't work properly wired makes it entirely broken to me. I am not willing to only play wirelessly with it. At some points I thought I recreated the issue on a regular GNU/Linux PC in a browser using gamepad-tester (last input press timestamp updating without me pressing anything), but I have not been able to do this reliably, so I'm not certain it was real.

Please let me know if there are any more useful details I can provide.

#### Steps for reproducing this issue:

1. Connect dock (official) or hub of some sort to Steam Deck
2. Plug Gulikit KK3 Max controller in via USB and set it to XInput Mode (if in Switch Mode, press the Mode button two times while powered on)
3. Use the controller a while or keep it in the main menu, may need to bump a stick or something occasionally if the issue doesn't happen within 10 minutes and the controller shuts itself off from inactivity

Super Hexagon has been a good test game for me because you only really need left and right on the DPad when playing, so the Steam menu opening is much more obviously not your own fault when it happens. I've also found that when it rains, it pours, and after the first instance I've had it happen 3 or 4 times within 10-15 minutes.

**Comment by Haxorzz** [comment]
In my experience using this on Linux (X-input android profile, nobara linux, bluetooth), I have yet to experience any phantom button presses and on Steam deck (Nintendo Switch profile, steamos preview, bluetooth) I also haven't had any phantom inputs that I can recall. Does this only happen when the steam deck has a hub connected to it and playing wired? Does this happen when using their wireless dongle?

**Comment by Soundtoxin** [comment]
As far as I've been able to tell, it only happens wired, plugged into a hub, yes. I have not experienced it in testing with the wireless dongle, but I'd prefer to play wired so I have not spent much time with the dongle or using bluetooth, just briefly trying them to test for the issue.

**Comment by broconut123** [comment]
Hey, the controller is basically a paperweight if you are running wired on linux. Just  by rotating the stick it will randomly press buttons and even change system volume. 

I have been in contact with Gulikit Support and they flat out said there are no plans to fix it for linux

![image](https://github.com/ValveSoftware/SteamOS/assets/92729707/81a1d9b0-dc43-4cb0-ac9f-37f3ed6ea92a)

**Comment by Soundtoxin** [comment]
@broconut123 Thanks for the info. That's unfortunate if true. Though, I wonder if it's a case of confusion over semantics. The support person you're dealing with may think "Linux" = unpopular/niche and not consider SteamOS to even be the same thing. When I had contacted support I got someone insisting they had done testing on the Steam Deck before release and that none of the engineers had run into this problem. That support person may also not realize that almost all controllers "just work" and it's more unusual that their new one doesn't, especially when the previous controller they made works fine.

For now I've just set my KK3 Max aside and I picked up an 8bitdo Pro 2, the new revision with hall effect sticks. I would like the KK3 Max to be fixed, but since I only care about using it wired it's mostly pretty useless to me right now.
I did do some testing on my PC running Guix System GNU/Linux where I didn't seem to hit my issue of the ghost presses, but it's probably just that the home button wasn't bound to anything there due to no Steam Big Picture or similar running. I was playing in Dolphin-emu with it, using Swaywm for my environment. I'm not sure what presses would've messed with volume but that sort of thing probably just isn't configured in my environment also. I don't have a graphical volume control or media keys or keyboard shortcuts set up for volume control.

I'm curious at what level they've screwed up their controller that it only happens in some modes/environments/operating systems. Would the correct fix be in firmware? Could something be done in the kernel to fix their issue for them? I've never run into something like this with a controller before.

**Comment by PickMeNow** [comment]
Using Gulikit KK3 MAX and I have ghost presses while playing games, sometimes it makes steam take screenshots while playing games, very frustrating :/  

With KK2 Pro this issue doesn't happen.

**Comment by broconut123** [comment]
> Using Gulikit KK3 MAX and I have ghost presses while playing games, sometimes it makes steam take screenshots while playing games, very frustrating :/
> 
> With KK2 Pro this issue doesn't happen.

![image](https://github.com/user-attachments/assets/7c3fbcff-22d1-46cc-aad7-9385323321f7)

**Comment by SilentPrayeCG** [comment]
Home screen randomly appear for me on Windows 11 as well, when kk3 max connected via cable. So it definitely not Linux's fault. And this isn't happen when it connected via dongle.

**Comment by Kimiblock** [comment]
Yeah... Using KK3 Pro here and dongle exists as a workaround. Switch mode + wired appears fine for a quick test though...

Arch Linux, it's not a steam fault either because GNOME Games and KDE Settings do see that ghost input.

**Comment by mrgreywater** [comment]
Adding
```c
	if (data[1] != 0x14) {
		return;
	}
```
to https://github.com/torvalds/linux/blob/master/drivers/input/joystick/xpad.c#L869
seems to fix it / filter out the bad data. No idea how compatible that is with other controllers though

**Comment by username-generic** [comment]
> Replying to https://github.com/ValveSoftware/SteamOS/issues/1451#issuecomment-4421103753

Applying this patch has resolved the issue for me as well. Please consider submitting a patch upstream, or poking the manufacturer.
